*: added a script mode

dgonyeo opened the pull request:
This commit adds a script mode to acbuild, whereby build scripts can be
written in an acbuild DSL and then run. When acbuild is run in this way,
acbuild will exec itself with the debug flag, a work-path flag, and the
contents of each line. This lives under the "script" subcommand, but
this subcommand will be implied if the run acbuild has the name
"acbuild-script" (this was necessary to support /usr/bin/env shebangs).
The apache example has been rewritten to be in the acbuild DSL.

With a DSL mode various enhancements are now possible (like per-command
build caches), but the immediate benefit is the fact that acbuild can
enforce cleanup of builds that use the script mode.

jonboulle commented:
Curious to get a review from @lucab on this

lucab commented on acbuild/acbuild.go:
```diff
@@ -239,6 +244,12 @@ func runWrapper(cf func(cmd *cobra.Command, args []string) (exit int)) func(cmd
+		args := os.Args[1:]
+		os.Args = []string{os.Args[0], "script"}
+		os.Args = append(os.Args, args...)
+	}
```
I think you are kind of fighting with multicall here. I was in need of something similar so I just split https://github.com/coreos/rkt/pull/2815 of my working branch. Do you think it could make things easier here?

sgotti replied:
Perhaps I'm missing something, but looks like this is related to cobra/pflags argument parsing (so cobra will execute `cmdScript`) than multicall.

lucab replied:
Yes, it is re-adjusting `os.Args` for cobra, but I think this is basically working-around/re-implementing multicall behavior.

sgotti replied:
@lucab ok, so you're suggesting to register the cmdScript as a multicall with name `acbuild-script`. Now I understand coreos/rkt#2815 :smile:

lucab replied:
(Sorry for having been cryptic) Yes! It probably won't solve the issue here (argv[0] adjusting for later re-exec) but I think it will help in better integrating with multicall machinery.

lucab commented on acbuild/script.go:
```diff
@@ -0,0 +1,168 @@
+func execScript(rawScript []byte) error {
+	script := strings.Split(string(rawScript), "\n")
+	for i := 0; i < len(script); i++ {
+		script[i] = trimComments(script[i])
```
I guess you can skip empty and newline-only lines here.

dgonyeo replied:
There's a check later on (right before it reexec's acbuild) to skip lines that are empty strings, and the `strings.TrimSpace` call at the end of `trimComments` should strip all whitespace, newlines or otherwise.

jonboulle replied:
for i, s := range script

dgonyeo replied:
I had forgotten that go lets you modify slices as you iterate over them. I'll change this.

lucab commented on acbuild/script.go:
```diff
@@ -0,0 +1,168 @@
+		script[i] = trimComments(script[i])
+	}
+
+	for _, line := range script {
```
Can this second line-by-line loop be squeezed into the first one above?

lucab commented on acbuild/script.go:
```diff
@@ -0,0 +1,168 @@
+	defer os.RemoveAll(tmpDir)
+	contextpath = tmpDir
+
+	for _, line := range script {
```
This is another line-by-line loop. From a quick glance it looks like all these three can be merged into a single one, or am I mislooking something?

dgonyeo replied:
I'll merge the first loop, but I want this to be separate from the other loop so I can search the whole input for run commands before execing anything. Otherwise I could get a few commands deep before discovering you're trying to use run as non-root.

lucab commented on acbuild/script.go:
```diff
@@ -0,0 +1,168 @@
+
+func execACBuild(workPath, line string) error {
+	args := []string{"--debug", "--work-path=" + workPath}
+	args = append(args, tokenizeLine(line)...)
```
I think it may be worth having some further validation here and check that argv[1] is a valid subcommand. I think you may also want to check for the "script in script" (is it a valid usecase?).

dgonyeo replied:
I do think that the "script in script" use case is valid. If a user wishes to separate out portions of their build process into separate files, it's trivial enough to support here. However having said that, I hadn't considered that this was possible and the current implementation resulted in the nested script both creating a new build directory and calling `acbuild end` when it completed, so I just pushed some changes that results in acbuild setting an environment variable and then looking for this variable when running a nested script.

I also don't want to do preemptive validation at this point, as some commands have aliases and all commands can be selected with a shorter version of their name (as in "dep" is a valid way to refer to the "dependency" command). Due to this, I think it's easier to just exec acbuild and let cobra do the checking.

lucab commented:
I'm typically not a fan of DSL, but I have to say that I'd really like having this scripting mode in acbuild! I've left some comments inline related to multicall usage, input validation and reducing runtime complexity.

dgonyeo commented:
I think this implementation of a DSL is a fairly sane way to add one, since the semantics are identical to whatever shell script you'd otherwise write, and if you want more power or control over a build you can still write a shell script just fine.

dgonyeo commented:
Ready for another review pass

jonboulle commented on acbuild/script.go:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	scriptName := args[0]
+	rawScript, err := ioutil.ReadFile(args[0])
```
scriptName

jonboulle commented on acbuild/script.go:
```diff
@@ -0,0 +1,188 @@
+
+	var tmpDir string
+	nestedScript := false
+	if inheritedPath := os.ExpandEnv("$" + nestedScriptEnvVar); inheritedPath != "" {
```
look down two functions in the godoc ;-) - https://golang.org/pkg/os/#Getenv

jonboulle commented on acbuild/script.go:
```diff
@@ -0,0 +1,188 @@
+	var tmpDir string
+	nestedScript := false
+	if inheritedPath := os.ExpandEnv("$" + nestedScriptEnvVar); inheritedPath != "" {
+		fmt.Printf("nested script! %s\n", inheritedPath)
```
!

jonboulle commented on acbuild/script.go:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	for _, line := range script {
+		if line != "" {
```
let's try remove some of this indenting?
if line == "" {
  continue
}

jonboulle commented on acbuild/script.go:
```diff
@@ -0,0 +1,188 @@
+	return strings.TrimSpace(line)
+}
+
+func tokenizeLine(line string) []string {
```
I'd like to see some nice table testing for this method. Tokenizer bugs are annoying.

jonboulle commented on acbuild/script.go:
```diff
@@ -0,0 +1,188 @@
+
+func tokenizeLine(line string) []string {
+	var tokens []string
+	var currentToken []rune
```
nit but bytes.Buffer (with WriteRune) would be more appropriate

jonboulle commented on acbuild/script.go:
```diff
@@ -0,0 +1,188 @@
+	isEscaped := false
+	for _, char := range line {
+		if isEscaped {
+			currentToken = append(currentToken, char)
```
What if someone tries to split a command across multiple lines?

jonboulle replied:
Some cases to throw at you:

```
# split command
run my-command arg1 \
     arg2
```

```
# oops, forgot a quote
add-anno my-anno 'some val
```

```
# oops, superfluous quote
label foo bar"
```

dgonyeo commented:
Addressed nits, added the ability to have multi-line commands with trailing `\` characters, and added a suite of test cases to try to hit the edge cases in the tokenizer and related functions.

jonboulle commented:
@squeed another one it might be nice to have you review

lucab commented on acbuild/script_test.go:
```diff
@@ -0,0 +1,251 @@
+	}
+}
+
+// no really guys, this language is _great_
```
:smile:

iaguis replied:
People always ask me, what about languages? I very quicky answer, we have the best languages, they're outstanding.

lucab commented:
No further concerns from my side.

squeed commented:
This testcase fails for `TestTrimComments`, should it?

```
testcase{
    "this is a \"comment in a #string\".",
    "this is a \"comment in a #string\".",
},
```

squeed commented:
It might be worth the extra effort to throw together a yacc grammar for this...

dgonyeo commented:
No, that shouldn't fail. I had forgotten about comments in strings. Since my comment parsing logic was run before the logic that's checking for strings, I removed `trimComments`, added its functionality to `tokenizeLine`, and added the cases `TestTrimComments` was checking for to `TestTokenizeLine`.

cdrage commented:
AWESOME work @dgonyeo ! I've been waiting for a functionality like this to be added. Just when I get back into looking at `acbuild` again in order to implement a DSL, I see this! Thanks for the awesome work!

I'd love to starting throwing out some examples for for acbuild scripts, one suggestion I have is the ability to use capitalization for each command.

Ex:
```
#!/usr/bin/env acbuild-script
BEGIN
SET-NAME example.com/apache

DEP add quay.io/coreos/alpine-sh


RUN -- apk update
RUN -- apk add apache2
RUN -- /bin/sh -c "echo 'ServerName localhost' >> /etc/apache2/httpd.conf"

PORT add http tcp 80

MOUNT add html /var/www/localhost/htdocs

SET-EXEC -- /bin/sh -c "chmod 755 / && /usr/sbin/httpd -D FOREGROUND"

WRITE --overwrite apache-latest-linux-amd64.aci
END
```

Also, is there the ability to remove the `--` for each command as it'll be implied in the scripting language?

dgonyeo commented:
Thanks :)

I have no issue with capitalizing the subcommand, acbuild can just make the first token of each line lowercase before running the command.

As far as removing the `--`, that exists due to the need to differentiate between flags to acbuild and flags to the command you're running. Right now, you could have a script with this line:

```
run --working-dir=/opt -- mycoolapp --some-option .
```

Without the `--`, cobra will try to parse `--some-option` as a flag to acbuild.

Maybe a way to remove it would be to have acbuild insert a `--` before the first argument that doesn't start with a `-`? This would mean that if you want to run a binary that starts with the `-` character it wouldn't work how you expect, but in that case you can just provide an absolute path to the binary.

cdrage commented:
@dgonyeo Thanks for replying!

Yeah, removing `--` would be _great_ as it'll simplify the experience when building an `.aci`. 

Although you are right. It'll create some large differentiation between the command line as well as the scripting language. 

Would the best way be to create documentation explaining it? Similar to other composable formats (I'm thinking kubernetes) as long as we have a decent amount of documentation flowing it'll be easier for users to switch to building via a DSL.

dgonyeo commented:
It is documented [here](https://github.com/appc/acbuild/blob/master/Documentation/subcommands/run.md#options-parsing), if you think the description there isn't sufficient I could always write a little more on the reasoning behind it.

And I'm comfortable with UX related diverges from the cli syntax. It should still be pretty clear the relation between the two syntaxes, and the `--` has been a common complaint.

dgonyeo commented:
Just pushed a commit that lowercases the first token of a line, and inserts a `--` before the first argument on a `run` command that's lacking a `--`.

squeed commented:
This seems reasonable to me.

A question: what are the security implications of running untrusted acbuild scripts?

dgonyeo commented:
Were I to hand someone a malicious acbuild script, a couple things come to mind:
- I could copy any files from your host and ship them off somewhere
- I could attempt to break out of systemd-nspawn or the chroot engine (the latter is probably easier), and if I did I'd have root on the box
- I could write an ACI to any path, potentially overwriting things you'd find important

So I'd highly recommend reviewing an acbuild script before running it. There's an issue from forever ago (here: https://github.com/appc/acbuild/issues/86) for running acbuild inside of rkt, and with the recent chroot engine this should now be much easier to implement, if someone wanted a sandboxed way to run an acbuild script.

squeed commented:
Ahh, all good points! Thanks!

dgonyeo commented:
Review ping?

jonboulle commented:
@squeed @lucab @alban

squeed commented:
👍 LGTM!

cdrage commented on examples/apache/build-apache.acb:
```diff
@@ -0,0 +1,28 @@
+dep add quay.io/coreos/alpine-sh
+
+# Install apache
+run -- apk update
```
doesn't `--` need to be removed (since you updated it to assume it when running `run` options, etc?)

cdrage replied:
ahhhh, saw your comment that said it will automatically add if it's missing, nvm! so this should work regardless.

dgonyeo replied:
Just double checked, this script runs fine.

lucab commented:
LGTM too.
